devito/ops/transformer.py

```diff
@@ -82,6 +85,104 @@ def to_ops_stencil(param, accesses):
     ]
 
 
+def create_ops_dat(f, name_to_ops_dat, block):
+    ndim = f.ndim - (1 if f.is_TimeFunction else 0)
+
+    dim = Array(
+        name=namespace['ops_dat_dim'](f.name),
+        dimensions=(DefaultDimension(name='dim', default_value=ndim),),
+        dtype=np.int32
+    )
+    base = Array(
+        name=namespace['ops_dat_base'](f.name),
+        dimensions=(DefaultDimension(name='base', default_value=ndim),),
+        dtype=np.int32
+    )
+    d_p = Array(
+        name=namespace['ops_dat_d_p'](f.name),
+        dimensions=(DefaultDimension(name='d_p', default_value=ndim),),
+        dtype=np.int32
+    )
+    d_m = Array(
+        name=namespace['ops_dat_d_m'](f.name),
+        dimensions=(DefaultDimension(name='d_m', default_value=ndim),),
+        dtype=np.int32
+    )
+
+    res = []
+    base_val = [Zero() for i in range(ndim)]
+
+    if f.is_TimeFunction:
+        time_pos = f._time_position
+        time_index = f.indices[time_pos]
+        time_dims = f.shape[time_pos]
+
+        dim_shape = sympify(f.shape[:time_pos] + f.shape[time_pos + 1:])
+        padding = f.padding[:time_pos] + f.padding[time_pos + 1:]
+        halo = f.halo[:time_pos] + f.halo[time_pos + 1:]
+        d_p_val = tuple(sympify([p[0] + h[0] for p, h in zip(padding, halo)]))
+        d_m_val = tuple(sympify([-(p[1] + h[1]) for p, h in zip(padding, halo)]))
+
+        ops_dat_array = Array(
+            name=namespace['ops_dat_name'](f.name),
+            dimensions=(DefaultDimension(name='dat', default_value=time_dims),),
+            dtype='ops_dat',
+        )
+
+        dat_decls = []
+        for i in range(time_dims):
```

Comment:
`for i in range(time_dims)`
So if there are a thousand time-steps this creates a thousand declarations?

Reply:
We need a declaration for each time stepping variables (`t0`, `t1`, etc.) because the time loop is not offloaded to OPS.

Reply:
`time_dims = f.shape[time_pos]`

That's one per time-steps, so again for 1000 time-steps that's 1000 declaration. The time dimension is not always a buffered dimension, it can be a normal dimension in which case this is gonna break.